File: Rust/Week1/Day4/day4/src/bin/ex5.rs
use std::{any::{Any, TypeId, type_name, type_name_of_val}, collections::HashMap};

    const BUFF_NAMES: [&str;2] = ["Shield", "Giant Stength"];
    const DEBUFF_NAMES: [&str;3] = ["Paralized", "Frozen", "Burning"];

enum Event<'a> {
    Damage {origin: &'a str, target: &'a str, amount: i32},
    Heal   {origin: &'a str, target: &'a str, amount: i32},
    Buff   {origin: &'a str, target: &'a str, name: &'a str},
    Debuff {origin: &'a str, target: &'a str, name: &'a str},
    Death  {origin: &'a str, target: &'a str, message: &'a str}
}

fn simulate_battle<'a>(
    party: &mut HashMap<&str, (i32, i32, &'a str)>,
    enemy_party: &mut HashMap<&str, (i32, i32, &'a str)>,
    logs: &mut Vec<Event<'a>>,
) {
    if let Some(stats) = party.get_mut("Astarion") {
        stats.0 += 1;
    }

    logs.push(Event::Heal {
        origin: "Shadowheart",
        target: "Astarion",
        amount: 1,
    });

    if let Some(stats) = party.get_mut("Astarion") {
        stats.2 = BUFF_NAMES[0];
    }

    logs.push(Event::Buff {
        origin: "Shadowheart",
        target: "Astarion",
        name: BUFF_NAMES[0],
    });
}


fn print_logs(logs:&Vec<Event<'_>>, stop_at_first_death: bool) {
    for event in logs.iter() {
        match event {
            Event::Damage {origin, target, amount } => {
                println!("{origin} dealt {amount} damage to {target}!");
            }
            Event::Heal   {origin, target, amount } => {
                println!("{origin} healed {amount} health points on {target}!");
            }
            Event::Buff   {origin, target, name } => {
                println!("{origin} buffed {target} with {name}!");
            }
            Event::Debuff {origin, target, name } => {
                println!("{origin} caused {name} on {target}!");
            }
            Event::Death  {origin, target, message } => {
                println!("{origin} has killed {target}! {target}: \"{message}\"");

                if stop_at_first_death {
                    println!("Logs stopped prematurely!");
                    break;
                }
            }
        }
    }
}

fn main() {
    let mut party = HashMap::from([
        ("Astarion",    (32, 40, "")),
        ("Shadowheart", (20, 45, "")),
        ("Karlach",     (55, 60, "")),
        ("Wyll",        (20, 50, "")),
        ("Lae'zel",     (47, 50, ""))
    ]);

    let mut enemy_party = HashMap::from([
        ("Goblin",     (40, 40, "")),
        ("Evil Druid", (45, 45, "")),
        ("Bear",       (80, 80, "")),
        ("Beholder",   (100, 100, "")),
        ("Wolf",       (47, 50, ""))
    ]);



    let mut log = vec![
        Event::Damage {origin: "Goblin", target: "Astarion", amount: 12 },
        Event::Heal   {origin: "Shadowheart", target: "Shadowheart", amount: 6 },
        Event::Buff   {origin: "Karlach", target: "Karlach", name: "Rage" },
        Event::Damage {origin: "Goblin", target: "Wyll", amount: 55 },
        Event::Debuff {origin: "Goblin", target: "Lae'zel", name: "Poison" },
        Event::Damage {origin: "Goblin", target: "Astarion", amount: 30 },
        Event::Death  {origin: "Goblin", target: "Astarion", message: "Astarion Death"}
    ];

    simulate_battle(&mut party, &mut enemy_party, &mut log);

    print_logs(&log, false);
}

File: CSharp/Week1/Day6/Day6/ex5.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Day6
{
    internal class ex5
    {   
        enum EventType { Damage = 0, Heal, Buff, Debuff, Death, Count}
        record EventLog{
            public required EventType Type;
            public required string Target;
            public required string Origin;
            public int Ammount;
            public required string Name;
        }

        struct Character
        {
            public String name;
            public int hp;
            public int maxHp;
            public Character(String name, int hp, int maxHp)
            {
                this.name = name;
                this.hp = hp;
                this.maxHp = maxHp;
            }
        }

        static void PrintLogs(ref List<EventLog> logs, bool stopAtFirstDeath = false)
        {
            var messages = new String[]
            {
                "{0} used {1} to attack {2} dealing {3} damage!",
                "{0} used {1} to heal {2} for {3} health points!",
                "{0} used {1} on {2}.",
                "{0} used {1} on {2}!",
                "{0} has killed {2}!"
            };
            foreach (EventLog log in logs)
            {
                Console.WriteLine(messages[(int)log.Type], log.Origin, log.Name, log.Target, log.Ammount);
            }
        }

        static void SimulateBattle(ref Random rand, ref List<EventLog> logs, ref List<Character> party, ref List<Character> enemyParty)
        {
            for (int i = 0; i < 10; i++)
            {
                EventType type = (EventType)rand.Next((int)EventType.Count);
                String target = party[rand.Next(party.Count)].name;
                String origin = enemyParty[rand.Next(enemyParty.Count)].name;
                int amount = rand.Next(10);

                String name = "Placeholder Skill Name";

                EventLog log = new EventLog { Type = type, Origin = origin, Target = target, Ammount = amount, Name = name };
                logs.Add(log);
            }
        }

        static void Main()
        {
            Random rand = new Random();

            List<Character> party = new List<Character>()
            {
                new Character("Astarion", 32, 40),
                new Character("Shadowheart", 20, 45),
                new Character("Karlach", 55, 60),
                new Character("Wyll", 1, 50),
                new Character("Lae'zel", 47, 50)
            };
            List<Character> enemyParty = new List<Character>()
            {
                new Character("Goblin", 32, 40),
                new Character("Gnol", 20, 45),
                new Character("Troll", 55, 60),
                new Character("Dire Wolf", 10, 50),
                new Character("Beholder", 47, 50)
            };

            var logs = new List<EventLog>();
            SimulateBattle(ref rand, ref logs, ref party, ref enemyParty);
            PrintLogs(ref logs, true);
        }
    }
}
